contrib/document_constants.py

```diff
@@ -0,0 +1,155 @@
+#!/usr/bin/env python3
+
+#
+# (C) 2021 ProFUSION Sistemas e Solucoes LTDA
+#
+# All files and artifacts in this repository are licensed under the
+# provisions of the license provided by the LICENSE file in this repository.
+#
+
+import argparse
+import sys
+from enum import Enum
+from textwrap import dedent
+from typing import Mapping, Sequence, Type, Union
+
+import vspec.model.constants as constants
+
+
+# NOTE: only used to fetch the auto-generated __doc__
+class GenericEnum(Enum):
+    placeholder = 1
+
+
+def create_table(
+    header: Sequence[str],
+    rows: Sequence[Sequence[str]],
+    size_rounding: int = 4,
+):
+    """Generate markdown table
+
+    It will measure all the row cells and their header to generate
+    a human friendly table (the text representation looks like a table).
+
+    To avoid too much resizing when the cells change, the sizes are rounded
+    using `size_rounding`.
+    """
+    def round(n):
+        remainder = n % size_rounding
+        if not remainder:
+            return n
+        return n + remainder
+
+    max_lengths = [round(len(h)) for h in header]
+    count = len(max_lengths)
+    for r in rows:
+        for i in range(count):
+            max_lengths[i] = max(max_lengths[i], round(len(r[i])))
+
+    fmt = "|"
+    for ml in max_lengths:
+        fmt += f" %-{ml}s |"
+
+    ret = [
+        fmt % header,
+        "|" + "|".join(f":{'-' * n}-" for n in max_lengths) + "|",
+    ]
+    for r in rows:
+        ret.append(fmt % r)
+    return "\n".join(ret)
+
+
+class DocumentEnum:
+    kind = Enum
+    cls: Enum
+    table_header = ("Symbol", "Value")
+
+    def __init__(self, cls: Enum) -> None:
+        self.cls = cls
+
+    @property
+    def name(self):
+        return self.cls.__name__
+
+    def __str__(self) -> str:
+        lines = []
+        if self.cls.__doc__ and self.cls.__doc__ != GenericEnum.__doc__:
+            doc = self.cls.__doc__.splitlines()
+            if len(doc) > 2 and doc[0] and not doc[1]:
+                lines.append(f"## {self.name} ({doc[0].strip()})")
+                doc.pop(0)
+                doc.pop(0)
+            else:
+                lines.append(f"## {self.name}")
+
+            lines.append(dedent("\n".join(doc)))
+            lines.append("")
+        else:
+            lines.append(f"## {self.name}")
+
+        rows = [
+            (f"`{m.name}`", m.value) for m in self.cls.__members__.values()
+        ]
+        lines.append(create_table(self.table_header, rows))
+        lines.append("")
+        return "\n".join(lines)
+
+
+DocumentGenerators = Union[Type[DocumentEnum]]
+Documents = Union[DocumentEnum]
```

Comment:
As far as I understand a Union type with only one choice is not useful.

So this
```
DocumentGenerators = Union[Type[DocumentEnum]]
Documents = Union[DocumentEnum]
```
is really the same as this (?)
```
DocumentGenerators = Type[DocumentEnum]
Documents = DocumentEnum
```

If you agree, I guess you could also use the original types directly in the expressions that follow instead of giving them a new name.  Unless the "renaming" is done just for easier understanding.  What's the idea here?

Reply:
yes, it's the same. The idea is that eventually we'll be documenting more than `Enum` and adding it there would be slightly more straightforward.

Say we add some document generator for other non-string classes, just `, OtherType` and it's done. But I don't mind just skipping the `Union[]` until we reach there

Reply:
see next